folly/io/async/AsyncSSLSocket.cpp

```diff
@@ -1218,6 +1246,13 @@ void AsyncSSLSocket::handleAccept() noexcept {
     SSL_set_msg_callback_arg(ssl_.get(), this);
   }
 
+#if FOLLY_OPENSSL_HAS_TLS13
+  if (!earlyDataPhaseComplete_ &&
+      handleReadEarlyData() != SSL_READ_EARLY_DATA_FINISH) {
```

Comment:
`handleReadEarlyData()` needs to return early to avoid potential buffer allocations if we weren't planning on handling any early data (`SSL_get_max_recv_early_data()` returns 0).

Reply:
According to OpenSSL [documentation](https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_get_max_early_data.html),

> The recv_max_early_data value also has an impact on early data that is accepted. The amount of data that is accepted will always be the lower of the max_early_data for the session and the recv_max_early_data setting for the server. If a client sends more data than this then the connection will abort.


Going with the above description, suppose the server sets the max_early_data = 1024 bytes and the same is notified to the client. Now, if recv_max_early_data = 0 and clients sends some early data whose size > 0, then the connection must abort. But the fact that client has sent some illegal amount of data will be recognised only if we call `SSL_read_early_data()` which will also require potential buffer allocation. If `handleReadEarlyData()` is returned early and `SSL_read_early_data()` is not invoked, then we cannot check if the client has sent some illegal amount of early data (in this case, size > 0 when recv_max_early_data = 0).
Thus, I think returning early might not be good idea in case of recv_max_early_data = 0.